### PR TITLE
Refactor FilterExec::statistics

### PR DESCRIPTION
**Pull request by berkaysynnada**
## Which issue does this PR close?

Closes #7553.

## Rationale for this change

This PR is a comprehensive refactor of the `statistics()` method of execs, based on the problem in the issue that will be closed. As mentioned in the issue, statistics function is causing an inefficiency in `FilterExec`, and the error propagation is broken. That is fixed, and the cases in which errors can arise are restricted and clearly defined. 

The main goal of this PR is to enhance DataFusion by introducing an enumeration called "Sharpness." This enumeration categorizes the level of information accuracy throughout the code. We're using it to categorize Statistics parameters more precisely and consistently across the codebase. 

## What changes are included in this PR?

1) `statistics()` method of exec's now returns a Result.
2) We've introduced the Sharpness enum along with its utility functions.
3) The option values of statistics parameters have been simplified, and the exactness of each field is now tracked separately.
4) Some optimizations on aggregations have been modified. Each aggregate now takes into consideration the corresponding statistics' exactness information.
5) Statistical calculations such as join output statistics are handled more wisely.
6) This change isn't directly related to this PR, but `NegativeExpr`'s are now supported in interval arithmetic.

## Are these changes tested?

Yes. Existing tests are adapted to Sharpness. New tests are added for the new functions.

### CODE REVIEW

File: datafusion/physical-plan/src/joins/utils.rs [hunk 1]
```diff
@@ -818,30 +819,45 @@ fn estimate_join_cardinality(
...
+        .column_statistics
+        .iter()
+        .zip(right_stats.column_statistics.iter())
+    {
+        // If there is no overlap in any of the join columns, that means the join
+        // itself is disjoint and the cardinality is 0. Though we can only assume
+        // this when the statistics are exact (since it is a very strong assumption).
+        if left_stat.min_value.get_value()? > right_stat.max_value.get_value()? {
+            return match (
+                left_stat.min_value.is_exact().unwrap(),
```
Comment (mustafasrepo) [inline datafusion/physical-plan/src/joins/utils.rs:838]: I think you should use unwrap_or(false)

File: datafusion/physical-plan/src/joins/utils.rs [hunk 2]
```diff
@@ -818,30 +819,45 @@ fn estimate_join_cardinality(
...
+                right_stat.max_value.is_exact().unwrap(),
+            ) {
+                (true, true) => Some(Sharpness::Exact(0)),
+                _ => Some(Sharpness::Inexact(0)),
+            };
+        }
+        if left_stat.max_value.get_value()? < right_stat.min_value.get_value()? {
+            return match (
+                left_stat.max_value.is_exact().unwrap(),
```
Comment (mustafasrepo) [inline datafusion/physical-plan/src/joins/utils.rs:847]: similar to above

File: datafusion/proto/src/physical_plan/from_proto.rs [hunk 3]
```diff
@@ -581,10 +583,92 @@ impl TryFrom<&protobuf::FileGroup> for Vec<PartitionedFile> {
...
         ColumnStatistics {
-            null_count: Some(cs.null_count as usize),
-            max_value: cs.max_value.as_ref().map(|m| m.try_into().unwrap()),
-            min_value: cs.min_value.as_ref().map(|m| m.try_into().unwrap()),
-            distinct_count: Some(cs.distinct_count as usize),
+            null_count: if cs.null_count.is_none() {
+                Sharpness::Absent
+            } else {
+                cs.null_count.clone().unwrap().into()
+            },
```
Comment (mustafasrepo) [inline datafusion/proto/src/physical_plan/from_proto.rs:590]: Write this section with if let do not use .unwrap

File: datafusion/proto/src/physical_plan/from_proto.rs [hunk 4]
```diff
@@ -581,10 +583,92 @@ impl TryFrom<&protobuf::FileGroup> for Vec<PartitionedFile> {
...
+            max_value: if cs.max_value.is_none() {
+                Sharpness::Absent
+            } else {
+                cs.max_value.clone().unwrap().into()
+            },
```
Comment (mustafasrepo) [inline datafusion/proto/src/physical_plan/from_proto.rs:595]: Similar to above cases

File: datafusion/proto/src/physical_plan/from_proto.rs [hunk 5]
```diff
@@ -581,10 +583,92 @@ impl TryFrom<&protobuf::FileGroup> for Vec<PartitionedFile> {
...
+            s_type
+        } else {
+            return Sharpness::Absent;
+        };
+        match sharpness_type {
+            protobuf::SharpnessInfo::Exact => {
+                if s.val.is_none() {
+                    Sharpness::Absent
+                } else if let Ok(ScalarValue::UInt64(Some(val))) =
+                    ScalarValue::try_from(&s.val.unwrap())
```
Comment (mustafasrepo) [inline datafusion/proto/src/physical_plan/from_proto.rs:622]: similar to above

File: datafusion/proto/src/physical_plan/from_proto.rs [hunk 6]
```diff
@@ -581,10 +583,92 @@ impl TryFrom<&protobuf::FileGroup> for Vec<PartitionedFile> {
...
+        let sharpness_type = if let Ok(s_type) = s.sharpness_info.try_into() {
+            s_type
+        } else {
+            return Sharpness::Absent;
+        };
+        match sharpness_type {
+            protobuf::SharpnessInfo::Exact => {
+                if s.val.is_none() {
+                    Sharpness::Absent
+                } else if let Ok(val) = ScalarValue::try_from(&s.val.unwrap()) {
```
Comment (mustafasrepo) [inline datafusion/proto/src/physical_plan/from_proto.rs:656]: similar to above

File: datafusion/proto/src/physical_plan/to_proto.rs [hunk 7]
```diff
@@ -63,7 +63,7 @@ use datafusion::physical_plan::expressions::{
 };
 use datafusion::physical_plan::{AggregateExpr, PhysicalExpr};
 
-use crate::protobuf::{self, physical_window_expr_node};
+use crate::protobuf::{self, physical_window_expr_node, scalar_value::Value};
```
Comment (mustafasrepo) [inline datafusion/proto/src/physical_plan/to_proto.rs:66]: Check for your usage of unwrap in to_proto and from_proto files

**Review by mustafasrepo — APPROVED** [review]
LGTM!.

File: datafusion/physical-plan/src/joins/utils.rs [hunk 8]
```diff
@@ -870,9 +890,18 @@ fn estimate_inner_join_cardinality(
 /// If distinct_count is available, uses it directly. If the column numeric, and
 /// has min/max values, then they might be used as a fallback option. Otherwise,
 /// returns None.
-fn max_distinct_count(num_rows: usize, stats: ColumnStatistics) -> Option<usize> {
-    match (stats.distinct_count, stats.max_value, stats.min_value) {
-        (Some(_), _, _) => stats.distinct_count,
+fn max_distinct_count(
```
Comment (ozankabak) [inline datafusion/physical-plan/src/joins/utils.rs:893]: Let's explore if we can remove the `get_int_range` function and use the cardinality functionality of the interval library here

**Review by ozankabak — APPROVED** [review]
Looks good, went through all of it. Left one comment, please take a look and we can discuss tomorrow. After addressing that point, let's finalize the PR body and send upstream.

**Pull request closed without merging**